common/src/main/java/com/wynntils/features/combat/AutoAttackFeature.java

```diff
@@ -0,0 +1,122 @@
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.bus.api.SubscribeEvent;
+
+@ConfigCategory(Category.COMBAT)
+public class AutoAttackFeature extends Feature {
+    private static final int TICKS_PER_ATTACK = 2;
+    private static final int SPELL_TIMEOUT_TICKS = 60;
+
+    private enum WeaponStatus {
+        UNKNOWN,
+        NOT_USABLE,
+        USABLE
+    }
+
+    private WeaponStatus weaponStatus = WeaponStatus.UNKNOWN;
+    private int lastSelectedSlot;
+    private int preventWrongCast = Integer.MIN_VALUE;
+
+    @SubscribeEvent
+    public void onChangeCarriedItemEvent(ChangeCarriedItemEvent event) {
+        weaponStatus = WeaponStatus.UNKNOWN;
+    }
+
+    @SubscribeEvent
+    public void onSetSlotEvent(SetSlotEvent.Post event) {
+        if (event.getSlot() == McUtils.inventory().selected) {
+            weaponStatus = WeaponStatus.UNKNOWN;
+        }
+    }
+
+    @SubscribeEvent
+    public void onSwing(ArmSwingEvent event) {
+        if (event.getActionContext() != ArmSwingContext.ATTACK_OR_START_BREAKING_BLOCK) return;
+        if (event.getHand() != InteractionHand.MAIN_HAND) return;
+        if (Models.Character.getClassType() != ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = McUtils.player().tickCount + SPELL_TIMEOUT_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onUseItem(UseItemEvent event) {
+        if (Models.Character.getClassType() == ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = McUtils.player().tickCount + SPELL_TIMEOUT_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onSpellCastCompleted(SpellEvent.Completed event) {
+        preventWrongCast = Integer.MIN_VALUE;
+    }
+
+    private boolean isHoldingUsableWeapon() {
+        ItemStack heldItem = McUtils.player().getItemInHand(InteractionHand.MAIN_HAND);
+        if (!ItemUtils.isWeapon(heldItem)) return false;
+        if (ItemUtils.getItemName(heldItem).contains("Unidentified")) return false;
+        for (StyledText lore : LoreUtils.getLore(heldItem)) {
+            if (lore.contains("✖")) return false;
+        }
+        return true;
+    }
+
+    @SubscribeEvent
+    public void onTick(TickEvent event) {
+        if (!Models.WorldState.onWorld()) return;
+
+        LocalPlayer player = McUtils.player();
+        int currentSelectedSlot = McUtils.inventory().selected;
+        if (currentSelectedSlot == lastSelectedSlot) {
+            if (preventWrongCast > player.tickCount) return;
+        } else {
+            lastSelectedSlot = currentSelectedSlot;
+            preventWrongCast = Integer.MIN_VALUE;
+        }
+
+        if (player.tickCount % TICKS_PER_ATTACK != 0) return;
+
+        if (weaponStatus == WeaponStatus.UNKNOWN) {
+            weaponStatus = isHoldingUsableWeapon() ? WeaponStatus.USABLE : WeaponStatus.NOT_USABLE;
+        }
```

Comment:
Do you think we need to even run this `== UNKNOWN` every tick? It's up to you since the check is cheap, but I think the chances of someone collecting a buff while holding an unusable weapon, then being able to use it, is incredibly low. Perhaps just do it once on hotbar switch?

Reply:
The costly check is just done once, if it is in an unknown state, then it will be either usable or nonusable. The comparison is cheap and can remain. I think this is a reasonable way to solve the problem of not doing too many checks, but still doing enough. If some pre-condition changes, we flag the status as being unknown, and then we re-check it once next tick.